include/seqan3/alignment/pairwise/policy/alignment_matrix_policy.hpp

```diff
@@ -0,0 +1,146 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::alignment_matrix_policy.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alignment/band/static_band.hpp>
+#include <seqan3/alignment/pairwise/detail/alignment_algorithm_cache.hpp>
+#include <seqan3/core/type_traits/basic.hpp>
+#include <seqan3/range/views/slice.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief Manages the alignment and score matrix.
+ *
+ */
+template <typename alignment_algorithm_t, typename score_matrix_type, typename trace_matrix_type>
+class alignment_matrix_policy
+{
+private:
+
+    //!\brief Allow alignment algorithm to instantiate this crtp base class.
+    friend alignment_algorithm_t;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alignment_matrix_policy() = default; //!< Defaulted.
+    constexpr alignment_matrix_policy(alignment_matrix_policy const &) = default; //!< Defaulted.
+    constexpr alignment_matrix_policy(alignment_matrix_policy &&) = default; //!< Defaulted.
+    constexpr alignment_matrix_policy & operator=(alignment_matrix_policy const &) = default; //!< Defaulted.
+    constexpr alignment_matrix_policy & operator=(alignment_matrix_policy &&) = default; //!< Defaulted.
+    ~alignment_matrix_policy() = default; //!< Defaulted.
+    //!}
+
+    /*!\brief Allocates the memory of the underlying matrices.
+     * \param[in] fst_range The first range (corresponds to horizontal dimension).
+     * \param[in] snd_range The second range (corresponds to vertical dimension).
+     *
+     * \details
+     *
+     * Initialises the underlying score and trace matrices and sets the respective matrix iterators to the begin of the
+     * corresponding matrix.
+     */
+    template <typename fst_range_t, typename snd_range_t>
+    constexpr void allocate_matrix(fst_range_t && fst_range, snd_range_t && snd_range)
```

Comment:
This has nothing to do with RAII. The class itself (the algorithm) does not know anything about the ranges it is invoked on. Thus the data storage is first allocated when the algorithm is invoked with the sequences. These of course follow the RAII principle, so that no memory will be leaked in case something goes wrong.

Reply:
Ok, I thought RAII is not only about avoiding the leak of memory but also to structure your data in a way that initialising it already allocates all necessary space (if possible).

Reply:
> Ok, I thought RAII is not only about avoiding the leak of memory but also to structure your data in a way that initialising it already allocates all necessary space (if possible).

It does initialises everything it needs when constructing the class. The sequences are not part of it. They are first given by the function call operator. Then the memory is initialised. Everything is RAII. Everything is as it ought to be.